keps/sig-api-machinery/4008-crd-ratcheting/README.md

```diff
@@ -323,11 +339,185 @@ myOtherField: newly added field
+would trigger a new `Duplicate key` error not previously seen for this object.
+
+These fields are not ratcheted because changing map keys or list type to a map alters 
+the structure of the schema, and old values cannot be correlated to new. 
+
+Additionally Server-Side-Apply does not tolerate updates that violate the map 
+keys constraint: these properties are unsafe to change without bumping CRD 
+version.
+
+### Atomic Lists and Maps
+
+Following Kubernetes convention that fields annotated `x-kubernetes-list-type`
+or `x-kubernetes-map-type` as `atomic` change as a unit, the ratcheting rules
+do not apply to values nested inside atomic lists and maps. But a DeepEqual 
+comparison on the entire object can still ratchet errors for subfields.
+
+### CEL Rules
+
+Non-transitional CEL Rules (x-kubernetes-validations that do not 
+make use of `oldSelf`) will be ratcheted, if the objects in scope to them 
+can be correlated, and old is equal to new.
+
+### Advanced Ratcheting
+
+In this KEP we propose a conservative catch-all ratcheting rule that should
+ provide a minimum level of ratcheting support for any validation rule in the 
+ schema.
+
+`DeepEqual(OldObject, NewObject)` requirement is very strong. But for some 
+validation rules it is too strict. Consider the `required` error. This
+validation is applied to the object owning the property. Thus, if a new field
+becomes required the object cannot have other fields added onto it or changed
+without also adding the newly required field.
+
+Kubernetes has precedent for using a weaker form of racheting when validation is
+strengthened:
+
+```go
+if !Validation(old) {
+  // Ignore errors caused by !Validation(new)
+}
+```
+
+An example of this is the structural schema requirement added when Kubernetes
+went GA ([source](https://github.com/kubernetes/kubernetes/blob/12dc19d46fb97cbbfeb1e12b8a10ff7ae73d9515/staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go#L1583-L1590)): 
+```go
+// requireStructuralSchema returns true if schemas specified must be structural
+func requireStructuralSchema(oldCRDSpec *apiextensions.CustomResourceDefinitionSpec) bool {
+	if oldCRDSpec != nil && specHasNonStructuralSchema(oldCRDSpec) {
+		// don't tighten validation on existing persisted data
+		return false
+	}
+	return true
+}
+```
+
+It is not possible to generically allow this weaker form of ratcheting on all
+types of validations - it may allow data that would have failed a validation
+on an older schema and break controller workflows.
+
+To enable schemas to add validations with more complicated ratcheting logic
+than a DeepEqual check (e.g. to allow a new `minLength` restriction to allow any
+value as long as it is getting closer the new minimum); we turn to CEL.
+
+#### Ratcheting Rules in CEL
+
+CRD Validation Rules currently have two types:
+
+- Normal Rule (Applies to `CREATE` and `UPDATE` requests)
+- Transition Rule (Applies to `UPDATE` requests only)
```

Comment:
I think what you're saying is that there's a difference between create and "add field during an update"?

Reply:
I think this is a great point and shows that basing off operations `CREATE`/`UPDATE` doesn't map perfectly well. With validations we really mean `new value` vs `changed value`. I think we can keep the KEP's intention as-is and defer the details to API review when changing the types. The important thing this KEP is adding is that we can have rules that only apply to new values.

Reply:
> I think this is a great point and shows that basing off operations CREATE/UPDATE doesn't map perfectly well. With validations we really mean new value vs changed value. I think we can keep the KEP's intention as-is and defer the details to API review when changing the types.

:100:

Reply:
Good catch @benluddy, I'd like to make sure this distinction is somehow captured in the KEP, then LGTM.

Reply:
Since we all agree `operations` is not what we want, I have replaced it with a new propoed boolean field `appliesOnCreate`, which causes a transition rule to be evaluated with `oldSelf = nil` on create or when old value could not be correlated. WDYT?